File: go/day20_part2_2024.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type point struct{ x, y int }

func main() {
	f, _ := os.Open("input.txt")
	defer f.Close()
	scanner := bufio.NewScanner(f)
	var grid []string
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	h, w := len(grid), len(grid[0])
	var S, E point
	walls := make([][]bool, h)
	for i := range walls {
		walls[i] = make([]bool, w)
	}
	var trackCells []point
	for i := 0; i < h; i++ {
		for j := 0; j < w; j++ {
			ch := grid[i][j]
			if ch == 'S' {
				S = point{i, j}
			} else if ch == 'E' {
				E = point{i, j}
			}
			if ch == '#' {
				walls[i][j] = true
			} else {
				trackCells = append(trackCells, point{i, j})
			}
		}
	}

	dirs := []point{{1,0},{-1,0},{0,1},{0,-1}}
	isTrack := func(x,y int)bool{
		return x>=0 && x<h && y>=0 && y<w && !walls[x][y]
	}

	normalDistFrom := func(start point)[][]int {
		dist := make([][]int,h)
		for i:=range dist {
			dist[i] = make([]int,w)
			for j:=range dist[i] {
				dist[i][j] = -1
			}
		}
		dist[start.x][start.y] = 0
		q := []point{start}
		for len(q)>0 {
			cur:=q[0];q=q[1:]
			for _,d:=range dirs {
				nx,ny:=cur.x+d.x,cur.y+d.y
				if nx<0||nx>=h||ny<0||ny>=w{continue}
				if walls[nx][ny]{continue}
				if dist[nx][ny]<0 {
					dist[nx][ny]=dist[cur.x][cur.y]+1
					q=append(q,point{nx,ny})
				}
			}
		}
		return dist
	}

	distFromS:=normalDistFrom(S)
	distFromE:=normalDistFrom(E)
	if distFromS[E.x][E.y]<0 {
		fmt.Println(0)
		return
	}
	normalCost := distFromS[E.x][E.y]

	// We allow cheats up to length 20 ignoring walls.
	// Cheats are uniquely identified by start and end positions.
	// We'll store best cost for each (start,end) pair.
	type pair struct {
		sx,sy,ex,ey int
	}
	cheats := make(map[pair]int)

	for _,startPos:=range trackCells {
		sd:=distFromS[startPos.x][startPos.y]
		if sd<0 {continue}

		// BFS up to 20 steps ignoring walls
		distC := make([][]int,h)
		for i:=range distC {
			distC[i]=make([]int,w)
			for j:=range distC[i] {
				distC[i][j]=-1
			}
		}
		distC[startPos.x][startPos.y]=0
		q:=[]point{startPos}

		for len(q)>0 {
			cur:=q[0];q=q[1:]
			steps:=distC[cur.x][cur.y]
			if steps==20 {continue}
			for _,d:=range dirs {
				nx,ny:=cur.x+d.x,cur.y+d.y
				if nx<0||nx>=h||ny<0||ny>=w{continue}
				// can go through walls here
				if distC[nx][ny]<0 {
					distC[nx][ny]=steps+1
					q=append(q,point{nx,ny})
				}
			}
		}

		// Now distC[x][y] = steps needed to reach (x,y) ignoring walls, or -1 if not reachable
		for x:=0;x<h;x++ {
			for y:=0;y<w;y++ {
				s:=distC[x][y]
				if s>0 && s<=20 && isTrack(x,y) {
					ed:=distFromE[x][y]
					if ed<0 {continue}
					cost := sd+s+ed
					if cost<normalCost {
						key:=pair{startPos.x,startPos.y,x,y}
						old,ok:=cheats[key]
						if !ok||cost<old{
							cheats[key]=cost
						}
					}
				}
			}
		}
	}

	count:=0
	for _,cost:=range cheats {
		saving:=normalCost - cost
		if saving>=100 {
			count++
		}
	}
	fmt.Println(count)
}

File: kotlin/day20_part2_2024.kt

import java.io.File
import java.util.LinkedList

data class Point(val x: Int, val y: Int)

fun main() {
    val grid = File("input.txt").readLines()
    val h = grid.size
    val w = grid[0].length
    var start = Point(0, 0)
    var end = Point(0, 0)
    val walls = Array(h) { BooleanArray(w) }
    val trackCells = mutableListOf<Point>()

    for (i in 0 until h) {
        for (j in 0 until w) {
            val ch = grid[i][j]
            if (ch == 'S') {
                start = Point(i, j)
            } else if (ch == 'E') {
                end = Point(i, j)
            }
            if (ch == '#') {
                walls[i][j] = true
            } else {
                trackCells.add(Point(i, j))
            }
        }
    }

    val dirs = listOf(Point(1, 0), Point(-1, 0), Point(0, 1), Point(0, -1))

    fun normalDistFrom(start: Point): Array<IntArray> {
        val dist = Array(h) { IntArray(w) { -1 } }
        dist[start.x][start.y] = 0
        val q = LinkedList<Point>()
        q.add(start)
        while (q.isNotEmpty()) {
            val cur = q.poll()
            for (d in dirs) {
                val nx = cur.x + d.x
                val ny = cur.y + d.y
                if (nx in 0 until h && ny in 0 until w && !walls[nx][ny] && dist[nx][ny] == -1) {
                    dist[nx][ny] = dist[cur.x][cur.y] + 1
                    q.add(Point(nx, ny))
                }
            }
        }
        return dist
    }

    val distFromS = normalDistFrom(start)
    val distFromE = normalDistFrom(end)

    if (distFromS[end.x][end.y] == -1) {
        println(0)
        return
    }

    val normalCost = distFromS[end.x][end.y]
    val cheats = mutableMapOf<Pair<Point, Point>, Int>()

    for (startPos in trackCells) {
        val sd = distFromS[startPos.x][startPos.y]
        if (sd == -1) continue

        val distC = Array(h) { IntArray(w) { -1 } }
        distC[startPos.x][startPos.y] = 0
        val q = LinkedList<Point>()
        q.add(startPos)

        while (q.isNotEmpty()) {
            val cur = q.poll()
            val steps = distC[cur.x][cur.y]
            if (steps == 20) continue
            for (d in dirs) {
                val nx = cur.x + d.x
                val ny = cur.y + d.y
                if (nx in 0 until h && ny in 0 until w && distC[nx][ny] == -1) {
                    distC[nx][ny] = steps + 1
                    q.add(Point(nx, ny))
                }
            }
        }

        for (x in 0 until h) {
            for (y in 0 until w) {
                val s = distC[x][y]
                if (s in 1..20 && !walls[x][y]) {
                    val ed = distFromE[x][y]
                    if (ed == -1) continue
                    val cost = sd + s + ed
                    if (cost < normalCost) {
                        val key = Pair(startPos, Point(x, y))
                        if (!cheats.containsKey(key) || cost < cheats[key]!!) {
                            cheats[key] = cost
                        }
                    }
                }
            }
        }
    }

    var count = 0
    for (cost in cheats.values) {
        if (normalCost - cost >= 100) {
            count++
        }
    }
    println(count)
}
